runtime/cudaq/qis/qarray.h

```diff
@@ -0,0 +1,83 @@
+/*************************************************************** -*- C++ -*- ***
+ * Copyright (c) 2022 - 2023 NVIDIA Corporation & Affiliates.                  *
+ * All rights reserved.                                                        *
+ *                                                                             *
+ * This source code and the accompanying materials are made available under    *
+ * the terms of the Apache License 2.0 which accompanies this distribution.    *
+ *******************************************************************************/
+
+#pragma once
+
+#include "qview.h"
+
+namespace cudaq {
+
+namespace details {
+/// qarray<N> for N < 1 should be a compile error
+template <std::size_t N>
+concept ValidQArraySize = N > 0;
+} // namespace details
+
+/// @brief A qarray is an owning, compile-time sized container for qudits.
+/// The semantics of the qarray follows that of a std::array for qudits. It
+/// is templated on the number of qudits contained and the number of levels for
+/// the held qudits.
+template <std::size_t N, std::size_t Levels = 2>
+  requires(details::ValidQArraySize<N>)
+class qarray {
+public:
+  /// @brief Useful typedef indicating the underlying qudit type
+  using value_type = qudit<Levels>;
+
+private:
+  /// @brief Reference to the held / owned array of qudits
+  std::array<value_type, N> qudits;
+
+public:
+  /// Nullary constructor
+  qarray() {}
+
+  /// @brief qarray cannot be copied
+  qarray(qarray const &) = delete;
```

Comment:
nit: consider `const` on left side here (assuming that is more consistent with other code).